Cpp_hands_on_programming/lib/pipe/Ipc_pipe.cc

```diff
@@ -35,55 +50,103 @@ void Ipc_pipe::send() {
 			throw(std::runtime_error("Ipc_pipe::send: readFile"));
 		}
 		// write to pipe
-		if ((bytes_write = write(fdp, buffer.data(), bytes_read)) == -1) {
-	    	throw(std::runtime_error("Ipc_pipe::send: write to pipe"));
-		}
+		t.startTimer();   
+		do {
+			errno = 0;
+			signal(SIGPIPE, SIG_IGN); // ignore SIGPIPE in case read end is closed 
+			bytes_write = write(fdp, buffer.data(), bytes_read);
+			if (bytes_write == -1) {
+				if (errno == EAGAIN) {
+					t.checkTimer("Pipe is full for too long. Errno: " + std::string(strerror(errno)));
+				} else {
+					throw(std::runtime_error("Ipc_pipe::send: write. Errno: " + std::string(strerror(errno))));
+				}
+			}
+		} while (errno == EAGAIN);
 		// compare read and written bytes 
 		if (bytes_read == bytes_write) {
 			total += bytes_read;
 		} else {
 			throw(std::runtime_error("Ipc_pipe::send: read write difference"));
 		}
-	}
-
-	close(fdp);
+	}  
+ 	close(fdp);
 	std::cout << "File delivered by pipe successfully, exiting the program.." << std::endl;
 }
 
 void Ipc_pipe::receive() {
     std::cout << "Starting pipeReceive.." << std::endl;
 	int fdp, bytes;
-	off_t fileSize(0), total(0);
+	off_t fileSize(0), total(0), sendFileSize(0);
+	bool headerSent(false);
 	std::vector<char> buffer(PIPE_BUF);
 	// open pipe 
-	while ((fdp = open(name.c_str(), O_RDONLY)) == -1) {
-		std::this_thread::sleep_for (std::chrono::seconds(1));
-	}
+	t.startTimer();
+	do {
+		errno = 0;
+		fdp = open(name.c_str(), O_RDONLY | O_NONBLOCK); 
+		t.checkTimer("Cannot connect to pipeSend. Errno: " + std::string(strerror(errno)), "Connecting pipeSend..");
+	} while(errno == ENOENT);
+
+	std::cout << "Connected!" << std::endl;
 	// Start read and write loop
-	while ((bytes = read(fdp, buffer.data(), PIPE_BUF)) != 0) {	
-		if (bytes == -1) {
-			throw(std::runtime_error("Ipc_pipe::receive: read"));
-		}
+	do {
+		t.startTimer();
+		do {
+			errno = 0;
+			bytes = read(fdp, buffer.data(), PIPE_BUF);
+			if (bytes == -1) {
+				if (errno == EAGAIN) {
+					t.checkTimer("Pipe is empty for too long. Errno: " + std::string(strerror(errno)));
+				} else {
+					throw(std::runtime_error("Ipc_pipe::receive: read. Errno: " + std::string(strerror(errno))));
+				}
+			}
+			else if (!headerSent) {
+				if (bytes == 0) {
+					t.checkTimer("Takes too long for write end to open");
+				} else {
+					auto sendFileName = std::string(buffer.begin(), buffer.begin() + bytes);
+					if ((sendFileSize = getFileSize(sendFileName)) == -1) {
```

Comment:
Is that not cheating to get the size of the file by getting the filename? :D

Reply:
And it could not work if the sender and the receiver are not launched from the same directory

Reply:
![image](https://user-images.githubusercontent.com/97879298/166454208-4ba0143e-82bf-4df3-b7ab-2570d32ab9ff.png)

Reply:
My intention is to use the file name to handle the case - same file name for sender and receiver

You are right, It does not work for different directory launch, I think this happen because of different cwd of send and receive, receive  could not refer to the send file using the same path sender. Good point, will fix

Reply:
You can check getcwd() command.